### PR TITLE
`Transport` & `Engine`: factor out `getcwd()` & `chdir()` for compatibility with upcoming async transport

### PR DESCRIPTION
**Pull request by khsrali**
`Transport.getcwd()` & `Transport.chdir()` methods used to making life easier by allowing basically all `Transport` methods to function with relative paths. 

However, for upcoming [async changes](https://github.com/aiidateam/aiida-core/pull/6079), this was problematic, as several calculations ended up writing and fetching from wrong directories. 

This PR, gets ride of the habit of setting *working directory* for an entire instance of `Transport` class in engine.
Only because in `aiida-core` several calculations use a same instance of the `Transport`, (to avoid opening many channels, etc.)

### CODE REVIEW

File: src/aiida/transports/plugins/local.py [hunk 1]
```diff
@@ -724,26 +727,40 @@ def _exec_command_internal(self, command, **kwargs):
 
         command = bash_commmand + escape_for_bash(command)
 
+        if workdir:
+            cwd = workdir
+        else:
+            warn_deprecation(
```
Comment (khsrali) [inline src/aiida/transports/plugins/local.py:733]: I don't need this, as it is being raised properly in `getcwd()` itself

File: src/aiida/transports/plugins/ssh.py [hunk 2]
```diff
@@ -1300,7 +1311,13 @@ def _exec_command_internal(self, command, combine_stderr=False, bufsize=-1):
         channel = self.sshclient.get_transport().open_session()
         channel.set_combine_stderr(combine_stderr)
 
-        if self.getcwd() is not None:
+        if workdir is not None:
+            command_to_execute = f'cd {workdir} &&  ( {command} )'
+        elif self.getcwd() is not None:
+            warn_deprecation(
```
Comment (khsrali) [inline src/aiida/transports/plugins/ssh.py:1317]: I don't need this, as it is being raised properly in `getcwd()` itself

File: src/aiida/engine/daemon/execmanager.py [hunk 3]
```diff
@@ -178,11 +173,11 @@ def upload_calculation(
             # Note: this will possibly overwrite files
             for root, dirnames, filenames in code.base.repository.walk():
                 # mkdir of root
-                transport.makedirs(str(root), ignore_existing=True)
+                transport.makedirs(workdir.join(root), ignore_existing=True)
 
                 # remotely mkdir first
                 for dirname in dirnames:
-                    transport.makedirs(str(root / dirname), ignore_existing=True)
+                    transport.makedirs(workdir.join(root, dirname), ignore_existing=True)
```
Comment (GeigerJ2) [inline src/aiida/engine/daemon/execmanager.py:180]: Also here not sure what's wrong with the previous implementation?

File: src/aiida/engine/processes/calcjobs/tasks.py [hunk 4]
```diff
@@ -253,7 +253,6 @@ async def task_monitor_job(
     async def do_monitor():
         with transport_queue.request_transport(authinfo) as request:
             transport = await cancellable.with_interrupt(request)
-            transport.chdir(node.get_remote_workdir())
```
Comment (GeigerJ2) [inline src/aiida/engine/processes/calcjobs/tasks.py:256]: Same here

File: src/aiida/orm/nodes/data/remote/base.py [hunk 5]
```diff
@@ -96,22 +93,15 @@ def listdir(self, relpath='.'):
...
-                if exception.errno in (2, 20):  # directory not existing or not a directory
-                    exc = OSError(
-                        f'The required remote folder {full_path} on {self.computer.label} does not exist, is not a '
-                        'directory or has been deleted.'
-                    )
-                    exc.errno = exception.errno
-                    raise exc from exception
-                else:
-                    raise
+            full_path = os.path.join(self.get_remote_path(), relpath)
```
Comment (GeigerJ2) [inline src/aiida/orm/nodes/data/remote/base.py:96]: Might be a scope creep here, and probably unnecesary for the changes of this PR, but we could, at some point, fully remove
`os.path` calls with `pathlib` as it's now the recommended way to handle file system paths.

File: src/aiida/schedulers/plugins/bash.py [hunk 6]
```diff
@@ -26,11 +26,12 @@ class BashCliScheduler(Scheduler, metaclass=abc.ABCMeta):
     def submit_job(self, working_directory: str, filename: str) -> str | ExitCode:
```
Comment (GeigerJ2) [inline src/aiida/schedulers/plugins/bash.py:26]: Just something I noted: In many places the variable is called `working_directory`, while in others it's `workdir` now.

File: tests/conftest.py [hunk 7]
```diff
@@ -202,6 +202,67 @@ def _generate_work_chain(entry_point, inputs=None):
     return _generate_work_chain
 
 
+@pytest.fixture
+def generate_calcjob_node():
```
Comment (GeigerJ2) [inline tests/conftest.py:206]: This fixture looks pretty familiar :D
Maybe we could merge it with `generate_calculation_node`, providing an additional input argument for when a `CalcJob` is
being created to avoid duplication? Not sure if there are any other `CalcJob` specific changes in this implementation?

File: tests/transports/test_all_plugins.py [hunk 8]
```diff
@@ -1305,8 +1305,7 @@ def test_asynchronous_execution(custom_transport):
                 tmpf.write(b'#!/bin/bash\nsleep 10\n')
                 tmpf.flush()
 
-                transport.chdir('/tmp')
-                transport.putfile(tmpf.name, script_fname)
+                transport.putfile(tmpf.name, os.path.join('/tmp', script_fname))
```
Comment (GeigerJ2) [inline tests/transports/test_all_plugins.py:1308]: Alos here `pathlib` as we go forward?

**Review by GeigerJ2 — CHANGES_REQUESTED** [review]
Thanks, @khsrali! I think it's definitely an improvement not to rely on `getcwd` and `chdir` anymore. Quite a lot of changes to unpack. Given that all tests are passing, I assume it works, so I'd approve. Though, maybe we can discuss the `StrPath` class again in person, I don't think I fully understand its use.

**Review by GeigerJ2 — COMMENTED** [review]
Thanks, @khsrali! I think it's definitely an improvement not to rely on `getcwd` and `chdir` anymore. Quite a lot of changes to unpack. Given that all tests are passing, I assume it works, so I'd approve. Though, maybe we can discuss the `StrPath` class again in person, I don't think I fully understand its use.

**Comment by unkcpz** [comment]
> Given that all tests are passing, I assume it works, so I'd approve. Though, maybe we can discuss the StrPath class again in person, I don't think I fully understand its use.

I did a quick look on the PR, I have the same question on this. Would be nice to just use `Pathlib` I think.

File: src/aiida/engine/daemon/execmanager.py [hunk 9]
```diff
@@ -33,11 +33,12 @@
 from aiida.orm.utils.log import get_dblogger_extra
 from aiida.repository.common import FileType
 from aiida.schedulers.datastructures import JobState
+from aiida.transports.util import StrPath
 
 if TYPE_CHECKING:
     from aiida.transports import Transport
 
-REMOTE_WORK_DIRECTORY_LOST_FOUND = 'lost+found'
+REMOTE_workdirECTORY_LOST_FOUND = 'lost+found'
```
[the remark this answers is not in the document]
Comment (khsrali) [reply src/aiida/engine/daemon/execmanager.py:41]: done

File: src/aiida/engine/processes/calcjobs/calcjob.py [hunk 10]
```diff
@@ -643,7 +643,6 @@ def _perform_dry_run(self):
         with LocalTransport() as transport:
             with SubmitTestFolder() as folder:
                 calc_info = self.presubmit(folder)
-                transport.chdir(folder.abspath)
```
[the remark this answers is not in the document]
Comment (khsrali) [reply src/aiida/engine/processes/calcjobs/calcjob.py:646]: yes, not needed. 
In the new implementation, instead of changing directory here, `upload_calculation` will take it from `folder.abspath` (in the case of `dry_run`)

File: src/aiida/transports/util.py [hunk 11]
```diff
@@ -9,12 +9,58 @@
 """General utilities for Transport classes."""
```
[the remark this answers is not in the document]
Comment (khsrali) [reply src/aiida/transports/util.py:9]: I agree with you, but I'm afraid it might be backward incompatible. Some plugin use that `copy_from_remote_to_remote` function from it.

File: src/aiida/engine/processes/calcjobs/tasks.py (hunk 4 above)
Comment (khsrali) [reply src/aiida/engine/processes/calcjobs/tasks.py:256]: yup, I had moved that to the only place it was used: `always_kill()`

**Comment by khsrali** [comment]
Thanks a lot @GeigerJ2 for the review. Please check the new changes.

> Though, maybe we can discuss the `StrPath` class again in person, I don't think I fully understand its use.

Yes! as we discussed, I mainly suggested that for aesthetic reasons, I didn't want to call for `str()` every time passing a path to any of `transport` methods as they only accept string, for now.

Anyways, now I changed to `pathlib`, at some point we need to improve the whole transport interface to accept and work with path-like-objects instead of  `str`, but that is out of scope of this PR.

File: tests/transports/test_all_plugins.py (hunk 8 above)
Comment (khsrali) [reply tests/transports/test_all_plugins.py:1308]: done

File: src/aiida/orm/nodes/data/remote/base.py (hunk 5 above)
Comment (khsrali) [reply src/aiida/orm/nodes/data/remote/base.py:96]: Agreed!

File: src/aiida/engine/daemon/execmanager.py (hunk 3 above)
Comment (khsrali) [reply src/aiida/engine/daemon/execmanager.py]: `root` could be a relative path to `workdir`

File: src/aiida/schedulers/plugins/bash.py (hunk 6 above)
Comment (khsrali) [reply src/aiida/schedulers/plugins/bash.py:26]: Actually that's the only place that the variable is called `working_directory`, you'll find more `workdir` in the src. :smile:

File: tests/conftest.py (hunk 7 above)
Comment (khsrali) [reply tests/conftest.py:206]: yeah.. we could do that as well, I just thought maybe this way makes sense as in future we could add more features for this fixture... but yeah idk

**Comment by khsrali** [comment]
Reminder for me: 
update this doc string of `exec_command_wait_bytes`:
        `Enforce the execution to be run from the pwd (as given by self.getcwd), if this is not None.`

**Comment by khsrali** [comment]
@GeigerJ2 can you have a final look at this? 
kinda need this to merge to avoid conflict hell with the new PR I'm going to make, soon. 
Cheers

File: src/aiida/engine/daemon/execmanager.py [hunk 12]
```diff
@@ -173,25 +171,25 @@ def upload_calculation(
...
-                    transport.makedirs(workdir.join(root, dirname), ignore_existing=True)
+                    transport.makedirs(str(workdir.joinpath(root, dirname)), ignore_existing=True)
 
                 # Note, once #2579 is implemented, use the `node.open` method instead of the named temporary file in
                 # combination with the new `Transport.put_object_from_filelike`
                 # Since the content of the node could potentially be binary, we read the raw bytes and pass them on
                 for filename in filenames:
                     with NamedTemporaryFile(mode='wb+') as handle:
-                        content = code.base.repository.get_object_content((pathlib.Path(root) / filename), mode='rb')
+                        content = code.base.repository.get_object_content((Path(root) / filename), mode='rb')
```
Comment (GeigerJ2) [inline src/aiida/engine/daemon/execmanager.py:185]:
```suggestion
                        content = code.base.repository.get_object_content(Path(root) / filename, mode='rb')
```
Probably also works without the brackets

File: tests/conftest.py (hunk 7 above)
Comment (GeigerJ2) [reply tests/conftest.py:206]: As agreed during in-person discussion, we'll remove the currently duplicated code (from the `generate_calculation_node` fixture which I needed for testing the dumping functionality), as it's not really needed here, and instead keep the `generate_calcjob_node` fixture at the minimum needed to test the changes of this PR.

**Review by GeigerJ2 — COMMENTED** [review]
Nice, `pathlib` for the win. Thanks for adapting the code, @khsrali. Just let's fix the fixtures, then I'll approve!

**Review by GeigerJ2 — APPROVED** [review]
~fml~ lgtm!
